### ISSUE TITLE
quorum

### ISSUE THREAD

**Issue by silvio-dizmo** [moby/moby#34069]
**Description**

**Steps to reproduce the issue:**
1.
2.
3.

**Describe the results you received:**

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
(paste your output here)
```

**Output of `docker info`:**

```
(paste your output here)
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by thaJeztah** [comment]
I guess this is a danger of a distributed storage in general; it's an interesting question though, basically: "what should happen if a node looses connectivity with the swarm"?

- keep tasks running (connectivity could be lost temporarily, and can be restored), at risk that there's temporarily more tasks than desired (but will reconcile once connection is re-established)
- kill the task, because there's no quorum (either the node is a manager, and it lost quorum, or it's a node that no longer is connected to a manger). This option should against killing tasks immediately (killing all tasks due to a "blip" in network connectivity may not be desirable)



Neither of the above will satisfy all use cases, so this should be a deliberate choice; perhaps a service option (`--on-swarm-disconnect=kill-tasks` - naming is hard 😅 )


ping @stevvooe @aaronlehmann @aluzzardi wondering if you have thoughts on this; also perhaps there's already a proven solution for this 👍

**Comment by aaronlehmann** [comment]
I agree with your analysis, but wanted to point out terminology issue:

> "what should happen if a node looses connectivity with the swarm"?
> [...]
> because there's no quorum (either the node is a manager, and it lost quorum, or it's a node that no longer is connected to a manger)

Meeting quorum means that more than half of the managers are communicating with each other. The swarm may still have a quorum even if there is a connectivity problem with one of the worker or manager nodes.

**Comment by stevvooe** [comment]
@thaJeztah @aaronlehmann Analysis looks good here!

The other thing to note is that it is hard to tell the difference between a restart, a quorum-loss and a loss of connectivity, so we generally err on keeping the services up. Generally, such a system that ensures that a resource isn't accessed twice would have to err on killing services, which might not be ideal for a lot of use cases.

Services that rely on shared access to a single-point resource, like a network volume, should generally use a controlled single point. For example, if data on the volume will get corrupted when mounted more than once, the volume should disallow that.

@silvio-dizmo Have you seen this in production or were asking about behavior in this scenario?

**Comment by silvio-dizmo** [comment]
Yes we have a handful of internal services that don't really need to scale but where it would be nice to have good availability. By writing a custom volume plugin, the locking could probably be done, but it would be nice if docker could just solve this problem.

The most convenient way for me would be to be able to set a timeout (maybe per service), after which it will be shutdown if the node can't reach a manager that still has quorum. Only after that, should the service be restarted on a different node. Maybe there should be two timeouts one for soft and one for hard shutdown. They could just be set to infinity by default. Which would agree with the current behavior.

**Comment by kmbulebu** [comment]
Here's something I came up with for handling Postgres in a Swarm, run as a single task, backed by shared NFS. I tested what would happen if the node with Postgres became network isolated and the Swarm reschedules the Postgres task. The result was bad. 

This solution uses a lock file with the container hostname (ID) to ensure that Postgres can't be rescheduled without a clean shutdown of the last instance. It IS NOT battle tested.

https://github.com/kmbulebu/postgres_container_fencing

**Comment by dgancho** [comment]
I had the same issue with swarm on windows hosts.
My Bot service was running on a node, which lost connectivity to the swarm. The swarm started a new instance on another node, while the isolated node was still running the service too. It ended up with two clone instances of the Bot, both responding to users.
Restarting the host windows machine fixed this (the node rejoined by itself)
